File: app/src/common/shared/org/mozilla/vrbrowser/ui/views/BookmarksView.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.vrbrowser.ui.views;

import android.animation.ObjectAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.mozilla.vrbrowser.R;
import org.mozilla.vrbrowser.browser.AccountsManager;
import org.mozilla.vrbrowser.browser.BookmarksStore;
import org.mozilla.vrbrowser.browser.SettingsStore;
import org.mozilla.vrbrowser.browser.engine.Session;
import org.mozilla.vrbrowser.browser.engine.SessionStore;
import org.mozilla.vrbrowser.databinding.BookmarksBinding;
import org.mozilla.vrbrowser.ui.adapters.Bookmark;
import org.mozilla.vrbrowser.ui.adapters.BookmarkAdapter;
import org.mozilla.vrbrowser.ui.adapters.CustomLinearLayoutManager;
import org.mozilla.vrbrowser.ui.callbacks.BookmarkItemCallback;
import org.mozilla.vrbrowser.ui.callbacks.BookmarksCallback;
import org.mozilla.vrbrowser.utils.UIThreadExecutor;

import java.util.ArrayList;
import java.util.List;

import mozilla.appservices.places.BookmarkRoot;
import mozilla.components.concept.storage.BookmarkNode;
import mozilla.components.concept.sync.AccountObserver;
import mozilla.components.concept.sync.AuthType;
import mozilla.components.concept.sync.OAuthAccount;
import mozilla.components.concept.sync.Profile;
import mozilla.components.service.fxa.SyncEngine;
import mozilla.components.service.fxa.sync.SyncReason;
import mozilla.components.service.fxa.sync.SyncStatusObserver;

public class BookmarksView extends FrameLayout implements BookmarksStore.BookmarkListener {

    private BookmarksBinding mBinding;
    private ObjectAnimator mSyncingAnimation;
    private AccountsManager mAccountManager;
    private BookmarkAdapter mBookmarkAdapter;
    private boolean mIgnoreNextListener;
    private boolean mIsSyncEnabled;
    private boolean mIsSignedIn;
    private ArrayList<BookmarksCallback> mBookmarksViewListeners;
    private CustomLinearLayoutManager mLayoutManager;

    public BookmarksView(Context aContext) {
        super(aContext);
        initialize(aContext);
    }

    public BookmarksView(Context aContext, AttributeSet aAttrs) {
        super(aContext, aAttrs);
        initialize(aContext);
    }

    public BookmarksView(Context aContext, AttributeSet aAttrs, int aDefStyle) {
        super(aContext, aAttrs, aDefStyle);
        initialize(aContext);
    }

    @SuppressLint("ClickableViewAccessibility")
    private void initialize(Context aContext) {
        LayoutInflater inflater = LayoutInflater.from(aContext);

        mBookmarksViewListeners = new ArrayList<>();

        // Inflate this data binding layout
        mBinding = DataBindingUtil.inflate(inflater, R.layout.bookmarks, this, true);
        mBinding.setCallback(mBookmarksCallback);
        mBookmarkAdapter = new BookmarkAdapter(mBookmarkItemCallback, aContext);
        mBinding.bookmarksList.setAdapter(mBookmarkAdapter);
        mBinding.bookmarksList.setOnTouchListener((v, event) -> {
            v.requestFocusFromTouch();
            return false;
        });
        mBinding.bookmarksList.setHasFixedSize(true);
        mBinding.bookmarksList.setItemViewCacheSize(20);
        mBinding.bookmarksList.setDrawingCacheEnabled(true);
        mBinding.bookmarksList.setDrawingCacheQuality(View.DRAWING_CACHE_QUALITY_HIGH);

        mLayoutManager = (CustomLinearLayoutManager) mBinding.bookmarksList.getLayoutManager();

        mBinding.setIsLoading(true);
        mBinding.executePendingBindings();

        Drawable[] drawables = mBinding.syncButton.getCompoundDrawables();
        mSyncingAnimation = ObjectAnimator.ofInt(drawables[0], "level", 0, 10000);
        mSyncingAnimation.setRepeatCount(ObjectAnimator.INFINITE);

        mAccountManager = SessionStore.get().getAccountsManager();
        mAccountManager.addAccountListener(mAccountListener);
        mAccountManager.addSyncListener(mSyncListener);

        mIsSyncEnabled = mAccountManager.isEngineEnabled(SyncEngine.Bookmarks.INSTANCE);

        updateBookmarks();
        SessionStore.get().getBookmarkStore().addListener(this);

        updateCurrentAccountState();

        setVisibility(GONE);

        setOnTouchListener((v, event) -> {
            v.requestFocusFromTouch();
            return false;
        });
    }

    public void onDestroy() {
        SessionStore.get().getBookmarkStore().removeListener(this);
        mAccountManager.removeAccountListener(mAccountListener);
        mAccountManager.removeSyncListener(mSyncListener);
    }

    private final BookmarkItemCallback mBookmarkItemCallback = new BookmarkItemCallback() {
        @Override
        public void onClick(@NonNull View view, @NonNull Bookmark item) {
            mBinding.bookmarksList.requestFocusFromTouch();

            Session session = SessionStore.get().getActiveSession();
            session.loadUri(item.getUrl());
        }

        @Override
        public void onDelete(@NonNull View view, @NonNull Bookmark item) {
            mBinding.bookmarksList.requestFocusFromTouch();

            mIgnoreNextListener = true;
            SessionStore.get().getBookmarkStore().deleteBookmarkById(item.getGuid());
            mBookmarkAdapter.removeItem(item);
            if (mBookmarkAdapter.itemCount() == 0) {
                mBinding.setIsEmpty(true);
                mBinding.setIsLoading(false);
                mBinding.executePendingBindings();
            }
        }

        @Override
        public void onMore(@NonNull View view, @NonNull Bookmark item) {
            mBinding.bookmarksList.requestFocusFromTouch();

            int rowPosition = mBookmarkAdapter.getItemPosition(item.getGuid());
            RecyclerView.ViewHolder row = mBinding.bookmarksList.findViewHolderForLayoutPosition(rowPosition);
            boolean isLastVisibleItem = false;
            if (mBinding.bookmarksList.getLayoutManager() instanceof LinearLayoutManager) {
                LinearLayoutManager layoutManager = (LinearLayoutManager) mBinding.bookmarksList.getLayoutManager();
                int lastVisibleItem = layoutManager.findLastCompletelyVisibleItemPosition();
                if (rowPosition == layoutManager.findLastVisibleItemPosition() && rowPosition != lastVisibleItem) {
                    isLastVisibleItem = true;
                }
            }

            mBinding.getCallback().onShowContextMenu(
                    row.itemView,
                    item,
                    isLastVisibleItem);
        }

        @Override
        public void onFolderOpened(@NonNull Bookmark item) {
            int position = mBookmarkAdapter.getItemPosition(item.getGuid());
            mLayoutManager.scrollToPositionWithOffset(position, 20);
        }
    };

    private BookmarksCallback mBookmarksCallback = new BookmarksCallback() {
        @Override
        public void onClearBookmarks(@NonNull View view) {
            mBookmarksViewListeners.forEach((listener) -> listener.onClearBookmarks(view));
        }

        @Override
        public void onSyncBookmarks(@NonNull View view) {
            switch(mAccountManager.getAccountStatus()) {
                case NEEDS_RECONNECT:
                case SIGNED_OUT:
                    mAccountManager.getAuthenticationUrlAsync().thenAcceptAsync((url) -> {
                        if (url != null) {
                            mAccountManager.setLoginOrigin(AccountsManager.LoginOrigin.BOOKMARKS);
                            SessionStore.get().getActiveStore().loadUri(url);
                        }
                    });
                    break;

                case SIGNED_IN:
                    mAccountManager.syncNowAsync(SyncReason.User.INSTANCE, false);

                    mBookmarksViewListeners.forEach((listener) -> listener.onSyncBookmarks(view));
                    break;

                default:
                    throw new IllegalStateException("Unexpected value: " + mAccountManager.getAccountStatus());
            }
        }

        @Override
        public void onShowContextMenu(@NonNull View view, Bookmark item, boolean isLastVisibleItem) {
            mBookmarksViewListeners.forEach((listener) -> listener.onShowContextMenu(view, item, isLastVisibleItem));
        }
    };

    public void addBookmarksListener(@NonNull BookmarksCallback listener) {
        if (!mBookmarksViewListeners.contains(listener)) {
            mBookmarksViewListeners.add(listener);
        }
    }

    public void removeBookmarksListener(@NonNull BookmarksCallback listener) {
        mBookmarksViewListeners.remove(listener);
    }

    private SyncStatusObserver mSyncListener = new SyncStatusObserver() {
        @Override
        public void onStarted() {
            if (mAccountManager.isEngineEnabled(SyncEngine.Bookmarks.INSTANCE)) {
                mBinding.syncButton.setEnabled(false);
                mBinding.syncButton.setHovered(false);
                mSyncingAnimation.setDuration(500);
                mSyncingAnimation.start();
            }
        }

        @Override
        public void onIdle() {
            mIsSyncEnabled = mAccountManager.isEngineEnabled(SyncEngine.Bookmarks.INSTANCE);
            if (mIsSyncEnabled) {
                mBinding.syncButton.setEnabled(true);
                mBinding.syncButton.setHovered(false);
                mSyncingAnimation.cancel();
            }
            updateUi();
        }

        @Override
        public void onError(@Nullable Exception e) {
            if (mAccountManager.isEngineEnabled(SyncEngine.Bookmarks.INSTANCE)) {
                mBinding.syncButton.setEnabled(true);
                mBinding.syncButton.setHovered(false);
                mSyncingAnimation.cancel();
                mBinding.syncDescription.setText(getContext().getString(R.string.fxa_account_last_no_synced));
            }
        }
    };

    private void updateCurrentAccountState() {
        switch(mAccountManager.getAccountStatus()) {
            case NEEDS_RECONNECT:
            case SIGNED_OUT:
                mIsSignedIn = false;
                updateUi();
                break;

            case SIGNED_IN:
                mIsSignedIn = true;
                updateUi();
                break;

            default:
                throw new IllegalStateException("Unexpected value: " + mAccountManager.getAccountStatus());
        }
    }

    private AccountObserver mAccountListener = new AccountObserver() {

        @Override
        public void onAuthenticated(@NotNull OAuthAccount oAuthAccount, @NotNull AuthType authType) {
            mIsSignedIn = true;
            updateUi();
        }

        @Override
        public void onProfileUpdated(@NotNull Profile profile) {
        }

        @Override
        public void onLoggedOut() {
            mIsSignedIn = false;
            updateBookmarks();
            updateUi();
        }

        @Override
        public void onAuthenticationProblems() {
            mIsSignedIn = false;
            updateUi();
        }
    };

    private void updateUi() {
        if (mIsSignedIn) {
            mBinding.syncButton.setText(R.string.bookmarks_sync);
            mBinding.syncDescription.setVisibility(VISIBLE);

            if (mIsSyncEnabled) {
                mBinding.syncButton.setEnabled(true);
                long lastSync = mAccountManager.getLastSync();
                if (lastSync == 0) {
                    mBinding.syncDescription.setText(getContext().getString(R.string.fxa_account_last_no_synced));

                } else {
                    long timeDiff = System.currentTimeMillis() - lastSync;
                    if (timeDiff < 60000) {
                        mBinding.syncDescription.setText(getContext().getString(R.string.fxa_account_last_synced_now));

                    } else {
                        mBinding.syncDescription.setText(getContext().getString(R.string.fxa_account_last_synced, timeDiff / 60000));
                    }
                }

            } else {
                mBinding.syncButton.setEnabled(false);
                mBinding.syncDescription.setVisibility(GONE);
            }

        } else {
            mBinding.syncButton.setEnabled(true);
            mBinding.syncButton.setText(R.string.fxa_account_sing_to_sync);
            mBinding.syncDescription.setVisibility(GONE);
        }
    }

    private void updateBookmarks() {
        SessionStore.get().getBookmarkStore().getTree(BookmarkRoot.Root.getId(), true).thenAcceptAsync(this::showBookmarks, new UIThreadExecutor());
    }

    private void showBookmarks(List<BookmarkNode> aBookmarks) {
        if (aBookmarks == null || aBookmarks.size() == 0) {
            mBinding.setIsEmpty(true);
            mBinding.setIsLoading(false);

        } else {
            mBinding.setIsEmpty(false);
            mBinding.setIsLoading(false);
            mBookmarkAdapter.setBookmarkList(aBookmarks);
        }
        mBinding.executePendingBindings();
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);

        double width = Math.ceil(getWidth()/getContext().getResources().getDisplayMetrics().density);
        int firstVisibleItem = ((LinearLayoutManager)mBinding.bookmarksList.getLayoutManager()).findFirstVisibleItemPosition();
        int lastVisibleItem = ((LinearLayoutManager)mBinding.bookmarksList.getLayoutManager()).findLastVisibleItemPosition();
        mBookmarkAdapter.setNarrow(width < SettingsStore.WINDOW_WIDTH_DEFAULT, firstVisibleItem, lastVisibleItem);
    }

    // BookmarksStore.BookmarksViewListener
    @Override
    public void onBookmarksUpdated() {
        if (mIgnoreNextListener) {
            mIgnoreNextListener = false;
            return;
        }
        updateBookmarks();
    }

    @Override
    public void onBookmarkAdded() {
        if (mIgnoreNextListener) {
            mIgnoreNextListener = false;
            return;
        }
        updateBookmarks();
    }
}

File: app/src/common/shared/org/mozilla/vrbrowser/ui/views/BookmarksView.kt
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.vrbrowser.ui.views

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView

import org.mozilla.geckoview.AllowOrDeny
import org.mozilla.geckoview.GeckoResult
import org.mozilla.geckoview.GeckoSession
import org.mozilla.geckoview.WebRequestError
import org.mozilla.vrbrowser.R
import org.mozilla.vrbrowser.audio.AudioEngine
import org.mozilla.vrbrowser.browser.BookmarksStore
import org.mozilla.vrbrowser.browser.SessionStore
import org.mozilla.vrbrowser.databinding.BookmarksBinding
import org.mozilla.vrbrowser.ui.adapters.BookmarkAdapter
import org.mozilla.vrbrowser.ui.callbacks.BookmarkClickCallback
import org.mozilla.vrbrowser.ui.widgets.BookmarkListener
import org.mozilla.vrbrowser.utils.UIThreadExecutor

import java.util.ArrayList
import androidx.databinding.DataBindingUtil
import mozilla.appservices.places.BookmarkRoot

import mozilla.components.concept.storage.BookmarkNode
import mozilla.components.concept.storage.BookmarkNodeType
import java.util.function.Consumer

class BookmarksView(context: Context) : FrameLayout(context), GeckoSession.NavigationDelegate, BookmarksStore.BookmarkListener {
    // Inflate this data binding layout
    private val mBinding: BookmarksBinding = DataBindingUtil.inflate(
        LayoutInflater.from(context), R.layout.bookmarks, this, true
    )
    private val mBookmarkListeners: MutableList<BookmarkListener> = ArrayList()
    private val mAudio: AudioEngine? = AudioEngine.fromContext(context)
    private var mIgnoreNextListener: Boolean = false

    private val backButton: UIButton
    private val currentFolderNameView: TextView

    private val defaultRoot = BookmarkRoot.Mobile.id
    private val defaultFolderTitleWhenAbsent = context.getString(R.string.bookmarks_title)

    private val rootTitles = BookmarksStore.rootTitles(context)
    private val folderNavigationStack = mutableListOf<FolderOnStack>()

    private data class FolderOnStack(val guid: String, val title: String)

    private val mBookmarkClickCallback = object : BookmarkClickCallback {
        override fun onClick(bookmark: BookmarkNode) {
            mAudio?.playSound(AudioEngine.Sound.CLICK)

            when (bookmark.type) {
                // Ignore clicks on separators.
                BookmarkNodeType.SEPARATOR -> return
                // Load regular bookmarks.
                BookmarkNodeType.ITEM -> SessionStore.get().loadUri(bookmark.url)
                // Handle folder clicks.
                BookmarkNodeType.FOLDER -> folderClicked(bookmark)
            }
        }

        override fun onDelete(bookmark: BookmarkNode) {
            mAudio?.playSound(AudioEngine.Sound.CLICK)

            mIgnoreNextListener = true
            SessionStore.get().bookmarkStore.deleteBookmarkById(bookmark.guid)
            mBookmarkAdapter.removeItem(bookmark)
            if (mBookmarkAdapter.itemCount() == 0) {
                mBinding.isEmpty = true
                mBinding.isLoading = false
                mBinding.executePendingBindings()
            }
        }
    }

    private val mBookmarkAdapter: BookmarkAdapter = BookmarkAdapter(mBookmarkClickCallback, context)

    init {
        SessionStore.get().addNavigationListener(this)

        mBinding.bookmarksList.adapter = mBookmarkAdapter
        mBinding.isLoading = true
        mBinding.executePendingBindings()
        syncBookmarks()
        SessionStore.get().bookmarkStore.addListener(this)

        currentFolderNameView = findViewById(R.id.currentFolderName)
        backButton = findViewById(R.id.backButton)
        backButton.setOnClickListener { onBackPressed() }

        visibility = View.GONE
    }

    fun addListeners(vararg listeners: BookmarkListener) {
        mBookmarkListeners.addAll(listeners.toList())
    }

    fun onDestroy() {
        mBookmarkListeners.clear()
        SessionStore.get().bookmarkStore.removeListener(this)
    }

    fun onBackPressed() {
        folderNavigationStack.removeAt(folderNavigationStack.lastIndex)
        syncBookmarks()
    }

    fun folderClicked(bookmark: BookmarkNode) {
        folderNavigationStack.add(FolderOnStack(
            bookmark.guid, bookmark.title ?: defaultFolderTitleWhenAbsent)
        )
        syncBookmarks()
    }

    private fun notifyBookmarksShown() {
        mBookmarkListeners.forEach { it.onBookmarksShown() }
    }

    private fun notifyBookmarksHidden() {
        mBookmarkListeners.forEach { it.onBookmarksHidden() }
    }


    private fun syncBookmarks() {
        val rootToOpen = if (folderNavigationStack.isEmpty()) {
            defaultRoot
        } else {
            folderNavigationStack.last().guid
        }
        SessionStore.get().bookmarkStore.getBookmarks(rootToOpen).thenAcceptAsync(Consumer {
            this.showBookmarks(it)
        }, UIThreadExecutor())
    }

    private fun showBookmarks(aBookmarks: List<BookmarkNode>?) {
        currentFolderNameView.text = if (folderNavigationStack.isEmpty()) {
            rootTitles[defaultRoot]
        } else {
            folderNavigationStack.last().title
        }

        if (folderNavigationStack.isEmpty()) {
            backButton.visibility = View.GONE
        } else {
            backButton.visibility = View.VISIBLE
        }

        mBinding.isLoading = false
        // Don't display "no bookmarks" view when we're in an empty folder.
        if (aBookmarks == null || aBookmarks.isEmpty() && folderNavigationStack.isEmpty()) {
            mBinding.isEmpty = true

        } else {
            mBinding.isEmpty = false
            mBookmarkAdapter.setBookmarkList(aBookmarks)
        }
        mBinding.executePendingBindings()
    }

    override fun setVisibility(visibility: Int) {
        super.setVisibility(visibility)

        if (visibility == View.VISIBLE) {
            notifyBookmarksShown()

        } else {
            notifyBookmarksHidden()
        }
    }

    // NavigationDelegate

    override fun onLocationChange(session: GeckoSession, url: String?) {
        if (visibility == View.VISIBLE && url != null && url != ABOUT_BLANK) {
            notifyBookmarksHidden()
        }
    }

    override fun onCanGoBack(session: GeckoSession, canGoBack: Boolean) {

    }

    override fun onCanGoForward(session: GeckoSession, canGoForward: Boolean) {

    }

    override fun onLoadRequest(session: GeckoSession, request: GeckoSession.NavigationDelegate.LoadRequest): GeckoResult<AllowOrDeny>? {
        return GeckoResult.ALLOW
    }

    override fun onNewSession(session: GeckoSession, uri: String): GeckoResult<GeckoSession>? {
        return null
    }

    override fun onLoadError(session: GeckoSession, uri: String?, error: WebRequestError): GeckoResult<String>? {
        return null
    }

    // BookmarksStore.BookmarkListener
    override fun onBookmarksUpdated() {
        if (mIgnoreNextListener) {
            mIgnoreNextListener = false
            return
        }
        syncBookmarks()
    }

    companion object {
        private const val ABOUT_BLANK = "about:blank"
    }
}
